scene/2d/camera_2d.cpp

```diff
@@ -33,183 +33,223 @@
-};
+	if (Engine::get_singleton()->is_editor_hint()) {
+		update(); // Will just be drawn
+		return;
+	}
 
-Vector2 Camera2D::get_zoom() const {
-	return zoom;
-};
+	if (!initialized) {
+		_update_position();
+	}
+
+	ERR_FAIL_COND_MSG(custom_viewport && !ObjectDB::get_instance(custom_viewport_id),
+			"Custom viewport does not match custom viewport id");
+	Transform2D xform = _get_camera_transform().affine_inverse();
+	viewport->set_canvas_transform(xform);
+	Point2 screen_offset = (anchor_mode == ANCHOR_MODE_DRAG_CENTER ? (screen_size * 0.5) : Point2());
+	get_tree()->call_group(group_name, "_camera_moved", xform, screen_offset);
+}
 
-Transform2D Camera2D::get_camera_transform() {
-	if (!get_tree()) {
-		return Transform2D();
+void Camera2D::_update_size() {
+	if (!is_inside_tree() || !viewport || !current) {
+		return;
 	}
 
-	ERR_FAIL_COND_V(custom_viewport && !ObjectDB::get_instance(custom_viewport_id), Transform2D());
+	ERR_FAIL_COND_MSG(custom_viewport && !ObjectDB::get_instance(custom_viewport_id),
+			"Custom viewport does not match custom viewport id");
 
-	Size2 screen_size = _get_camera_screen_size();
+	if (anchor_mode == ANCHOR_MODE_DRAG_CENTER) {
+		current_position += screen_size * 0.5;
+		target_position += screen_size * 0.5;
+	}
+	screen_size = viewport->get_visible_rect().size * zoom_scale;
+	if (anchor_mode == ANCHOR_MODE_DRAG_CENTER) {
+		current_position -= screen_size * 0.5;
+		target_position -= screen_size * 0.5;
+	}
+	_update_position();
+}
 
-	Point2 new_camera_pos = get_global_position();
-	Point2 ret_camera_pos;
+void Camera2D::_update_position() {
+	if (!is_inside_tree() || !current) {
+		return;
+	}
 
-	if (!first) {
+	target_position = get_global_position();
+	if (!initialized) {
+		current_position = target_position;
 		if (anchor_mode == ANCHOR_MODE_DRAG_CENTER) {
-			if (drag_horizontal_enabled && !Engine::get_singleton()->is_editor_hint() && !drag_horizontal_offset_changed) {
-				camera_pos.x = MIN(camera_pos.x, (new_camera_pos.x + screen_size.x * 0.5 * zoom_scale.x * drag_margin[SIDE_LEFT]));
-				camera_pos.x = MAX(camera_pos.x, (new_camera_pos.x - screen_size.x * 0.5 * zoom_scale.x * drag_margin[SIDE_RIGHT]));
-			} else {
-				if (drag_horizontal_offset < 0) {
-					camera_pos.x = new_camera_pos.x + screen_size.x * 0.5 * drag_margin[SIDE_RIGHT] * drag_horizontal_offset;
-				} else {
-					camera_pos.x = new_camera_pos.x + screen_size.x * 0.5 * drag_margin[SIDE_LEFT] * drag_horizontal_offset;
-				}
+			current_position -= screen_size * 0.5;
+		}
+		initialized = true;
+	}
 
-				drag_horizontal_offset_changed = false;
+	if (anchor_mode == ANCHOR_MODE_DRAG_CENTER) {
+		Vector2 half_screen_size = screen_size * 0.5;
+		Point2 current_center = current_position + half_screen_size;
+		if (drag_horizontal_offset_changed) {
+			if (drag_horizontal_offset > 0) {
+				target_position.x += drag_horizontal_offset * drag_margin[SIDE_RIGHT] * half_screen_size.x;
+			} else { // drag_horizontal_offset < 0
```

Comment:
It's actually `<=`.

Reply:
Technically yes. I've updated it so it's accurate.